test/extensions/filters/http/jwt_authn/authenticator_test.cc

```diff
@@ -0,0 +1,468 @@
+    "h6nqKXcPNaRx9lOaRWg2PkE6ySNoyju7rNfunXYtVxPuUIkl0KMq3WXWRb_cb8a_Z"
+    "EprqSZUzi_ZzzYzqBNVhIJujcNWij7JRra2sXXiSAfKjtxHQoxrX8n4V1ySWJ3_1T"
+    "H_cJcdfS_RKP7YgXRWC0L16PNF5K7iqRqmjKALNe83ZFnFIw";
+
+} // namespace
+
+class MockAuthenticatorCallbacks : public Authenticator::Callbacks {
+public:
+  MOCK_METHOD1(onComplete, void(const Status& status));
+};
+
+class AuthenticatorTest : public ::testing::Test {
+public:
+  void SetUp() {
+    MessageUtil::loadFromYaml(ExampleConfig, proto_config_);
+    CreateAuthenticator();
+  }
+
+  void CreateAuthenticator() {
+    filter_config_ = ::std::make_shared<FilterConfig>(proto_config_, "", mock_factory_ctx_);
+    auth_ = Authenticator::create(filter_config_);
+  }
+
+  JwtAuthentication proto_config_;
+  FilterConfigSharedPtr filter_config_;
+  std::unique_ptr<Authenticator> auth_;
+  NiceMock<Server::Configuration::MockFactoryContext> mock_factory_ctx_;
+  MockAuthenticatorCallbacks mock_cb_;
+};
+
+// A mock HTTP upstream with response body.
+class MockUpstream {
+public:
+  MockUpstream(Upstream::MockClusterManager& mock_cm, const std::string& response_body)
+      : request_(&mock_cm.async_client_), response_body_(response_body) {
+    ON_CALL(mock_cm.async_client_, send_(_, _, _))
+        .WillByDefault(Invoke([this](Http::MessagePtr&, Http::AsyncClient::Callbacks& cb,
+                                     const absl::optional<std::chrono::milliseconds>&)
+                                  -> Http::AsyncClient::Request* {
+          Http::MessagePtr response_message(new Http::ResponseMessageImpl(
+              Http::HeaderMapPtr{new Http::TestHeaderMapImpl{{":status", "200"}}}));
+          response_message->body().reset(new Buffer::OwnedImpl(response_body_));
+          cb.onSuccess(std::move(response_message));
+          called_count_++;
+          return &request_;
+        }));
+  }
+
+  int called_count() const { return called_count_; }
+
+private:
+  Http::MockAsyncClientRequest request_;
+  std::string response_body_;
+  int called_count_{};
+};
+
+TEST_F(AuthenticatorTest, TestOkJWTandCache) {
+  MockUpstream mock_pubkey(mock_factory_ctx_.cluster_manager_, PublicKey);
+
+  // Test OK pubkey and its cache
+  for (int i = 0; i < 10; i++) {
+    auto headers = Http::TestHeaderMapImpl{{"Authorization", "Bearer " + GoodToken}};
+
+    MockAuthenticatorCallbacks mock_cb;
+    EXPECT_CALL(mock_cb, onComplete(_)).WillOnce(Invoke([](const Status& status) {
+      ASSERT_EQ(status, Status::Ok);
+    }));
+
+    auth_->verify(headers, &mock_cb);
+
+    EXPECT_EQ(headers.get_("sec-istio-auth-userinfo"),
+              "eyJpc3MiOiJodHRwczovL2V4YW1wbGUuY29tIiwic3ViIjoidGVzdEBleGFtcG"
+              "xlLmNvbSIsImV4cCI6MjAwMTAwMTAwMSwiYXVkIjoiZXhhbXBsZV9zZXJ2"
+              "aWNlIn0");
+    // Verify the token is removed.
+    EXPECT_FALSE(headers.Authorization());
+  }
+
+  EXPECT_EQ(mock_pubkey.called_count(), 1);
+}
+
+TEST_F(AuthenticatorTest, TestOkJWTPubkeyNoAlg) {
```

Comment:
Please add a one-liner `// This test validates foo and bar.` above each `TEST_F`.